oak_functions/loader/src/server.rs

```diff
@@ -461,19 +452,42 @@ impl wasmi::Externals for WasmState {
                     .read_extension_args(request_ptr, request_len)
                     .map_err(|err| {
                         self.log_error(&format!(
-                            "Handle {:?}: Unable to read input from guest memory: {:?}",
-                            extension.get_handle(),
+                            "Unable to read input from guest memory: {:?}",
                             err
                         ));
                         OakStatus::ErrInvalidArgs
                     });
 
+                let extension = &mut self
+                    .extensions_indices
+                    .get_mut(&index)
+                    .unwrap_or_else(|| panic!("Unimplemented function at {}", index));
+
                 let result = match request {
-                    Ok(request) => extension.invoke(self, args, request)?,
+                    Ok(request) => {
+                        let response = extension.invoke(request);
+                        match response {
+                            Ok(response) => {
+                                if !response.is_empty() {
+                                    // Careful: We assume every ABI call which returns a non-emtpy
+                                    // response provides these arguments
+                                    // (which is true).
+                                    let response_ptr_ptr: AbiPointer = args.nth_checked(2)?;
+                                    let response_len_ptr: AbiPointer = args.nth_checked(3)?;
+                                    self.write_extension_result(
+                                        response,
+                                        response_ptr_ptr,
+                                        response_len_ptr,
+                                    )
+                                } else {
+                                    Ok(())
+                                }
+                            }
+                            Err(err) => Err(err),
```

Comment:
Whenever you are passing erros unmodified, it probably means you could have shortcut the error via the `?` operator first, or at least you could use `Result::map` to only deal with the ok case.

Reply:
This is a bit tricky with the [`from_oak_status_result` before return](https://github.com/project-oak/oak/blob/ac0c134a93153e428fb6b4129c766cf28b1b9c02/oak_functions/loader/src/server.rs#L477). There is some magic happening here, which I hope to simplify later on.